### PR TITLE
Persistent ContainerDisks

### LINKED ISSUE #3638: copy on write volumes as delta for read only volumes.
/kind enhancement

The feature request here is to facilitate copy on write backing volumes for read-only volumes.

We already have a feature called an "ephemeralVolume" which allows us to attach a read-only PVC to a VMI, and automatically generate an ephemeral copy on write backing disk for that read only disk. However, that cow disk is completely ephemeral and only exists for the lifetime of the VMI. This means it can't be used to persistently store data that will survive a VM restart.

For a persistent cow disk, we need the ability to assign a PVC as a backing volume for a read only volume. I think we're looking at a new volume type to achieve this. Perhaps something like a ```copyOnWriteVolume```

The struct could look something like this.

```
type CopyOnWriteVolume struct {
    // BaseVolume is the read only source volume
    BaseVolume VolumeSource `json:"baseVolume"`

    // BaseVolumeSHA256 is the optional sha256 hash of the base volume used to validate the contents haven't mutated.
    BaseVolumeSHA256 string `json:"baseVolumeSHA256"`

    // PersistentVolumeClaim is the pvc to store the copy on write delta file
    PersistentVolumeClaim VolumeSource `json:"persistentVolumeClaim"`
}
```

## example VMI with copyOnWriteVolume with containerDisk

In this example, we use a containerDisk as the backing volume, and the PVC named `my-pvc` to maintain the cow delta file. 

```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: vm-cirros
  name: vm-cirros
spec:
  running: false
  template:
    metadata:
      labels:
        kubevirt.io/vm: vm-cirros
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: my-cow-volume
        resources:
          requests:
            memory: 64M
      volumes:
      - copyOnWriteVolume:
        name: my-cow-volume
        baseVolume:
            containerDisk:
                image: kubevirt/fedora32@sha123123123
        persistentVolumeClaim:
            name: my-pvc
```

## example VMI with copyOnWriteVolume with read only PVC

```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  labels:
    kubevirt.io/vm: vm-cirros
  name: vm-cirros
spec:
  running: false
  template:
    metadata:
      labels:
        kubevirt.io/vm: vm-cirros
    spec:
      domain:
        devices:
          disks:
          - disk:
              bus: virtio
            name: my-cow-volume
        resources:
          requests:
            memory: 64M
      volumes:
      - copyOnWriteVolume:
        name: my-cow-volume
        baseVolume:
            persistentVolumeClaim:
                name: my-golden-read-only-image
        persistentVolumeClaim:
            name: my-delta-pvc
```

## SHA validation of the base volume

We can back propagate a sha256 hash of the containerDisks contents to the VMI's status field and ultimately back to the VM's status in order to capture the sha256 hash the first time the VM runs. This hash could then be used the next time the VMI is executed to guarantee that the contents align as expected and that the containerDisk did not mutate.

This back propagating of a sha is likely most useful when we consider a read only PVC as the base volume though. I think for containerDisks as the base image, we should simply require the container image's sha to be used instead of a floating tag.

## SnapShot considerations

Snapshotting the copyOnWrite volume's delta PVC should be possible just like a normal PVC snapshot. In this case we'd simply be copying the delta PVC and ensuring we couple that PVC with the correct base volume during restore. The sha validation will guarantee we catch any issues that might occur if the source mutates.

## Usage Dangers.

There's only so much we can do to protect users from messing this logic up. If a user mutates the source volume behind the scenes, that's not something we can control. The best we can do is catch when this occurs using a hash. For container disks, I think it would be wise to require the usage of a sha hash reference instead of a floating tag. For PVCs we'll have to depend on our own sort of hash validation. 

# Implementation

## phase 1: container disks + cow

Decide what the API should look like and limit copy on write persistent storage to a single volume source (containerDisks). Since container disks use container images, we already have a solution to the sha issue. We simply require containerDisks with a cow delta file to be referenced using the image sha and not a floating tag.

## phase 2. PVCs + cow

Once we feel comfortable with the api and have proven it for container disks, we can approach the more complicated PVC use case. This will require expanding on what we did in phase 1 to provide some sort of sha validation for read only disks on PVCs.

### PR DESCRIPTION
**Pull request by davidvossel**
WIP. lets come to an agreement on the API, then I'll lock this down with func/unit tests. 

This PR adds a new, `CopyOnWrite`, field to ContainerDisks which makes it possible for ContainerDisks to be used with persistence across restarts. The API is structured in a way that allows expanding (in the future) this CopyOnWrite volume type to backing files that exist on other volume types other than ContainerDisks as well.

# Overview

The new CopyOnWrite field allows ContainerDisks to act as a backing file for a qcow2 image on a PVC. The workflow here around creating the qcow2 and validating the backing file is all automated. All a user needs to do in order to use a persistent containerDisk is to supply the VM/VMI with an empty PVC for the qcow2 file to live on.

# API Examples

Today, users can define an ephemeral ContainerDisk volume with the following volumes entry on their VM/VMI

```
  volumes:
  - containerDisk:
      image: kubevirt/fedora-cloud-container-disk-demo:devel
    name: containerdisk
```

With the new ```CopyOnWrite``` field, that same ContainerDisk can be used in a persistent way by assigning a PVC to accompany the ContainerDisk. All writes occur on the PVC, and the ContainerDisk is never mutated.

```
  volumes:
  - containerDisk:
      image: kubevirt/fedora-cloud-container-disk-demo:devel
      copyOnWrite:
        persistentVolumeClaim:
          claimName: my-pvc
    name: containerdisk
```

# TODO

- [ ] unit / functional tests
- [ ] user guide
- [ ] add feature gate
- [ ] add validation webhook logic to prevent using containers with `latest`. Should require tag or sha. 

Fixes #3638

**Release note**:
```release-note
Adds ability to have persistent ContainerDisk volumes with the new CopyOnWrite volume type
```

### CODE REVIEW

File: pkg/container-disk/container-disk.go [hunk 1]
```diff
@@ -160,11 +162,20 @@ func GenerateContainers(vmi *v1.VirtualMachineInstance, podVolumeName string, bi
 
 	// Make VirtualMachineInstance Image Wrapper Containers
 	for index, volume := range vmi.Spec.Volumes {
+
```
Comment (viorelcanja) [inline pkg/container-disk/container-disk.go:165]: The pattern of assigning a containerdisk vad based on the volume type (containerdisk or copyonwrite) appears 6 times. The code should be extracted in a function (something like GetReferencedContainerDisk). The same could be done for the referenced pvc for symmetry.

Comment (davidvossel) [reply pkg/container-disk/container-disk.go:165]: yep, great idea. i'll make sure to clean this up

**Comment by vladikr** [comment]
I'm just wondering if you've considered not to change the containerDisk API and introduce the `copyOnWrite` to be part of it? Something like the following:

```
volumes:
  - containerDisk:
      image: kubevirt/fedora-cloud-container-disk-demo:devel
      - copyOnWrite:
        persistentVolumeClaim:
          claimName: my-pvc
    name: containerdisk
```

**Comment by davidvossel** [comment]
> I'm just wondering if you've considered not to change the containerDisk API and introduce the copyOnWrite to be part of it? Something like the following:
>
> volumes:
>  - containerDisk:
>      image: kubevirt/fedora-cloud-container-disk-demo:devel
>      - copyOnWrite:
>        persistentVolumeClaim:
>          claimName: my-pvc
>    name: containerdisk

Interesting. I hadn't considered that. I actually like this way more than what I have now.

In the future we could even do the same thing for PVCs if we wanted.

```
 volumes:
  - persistentVolumeClaim:
      claimName: my-image-pvc
      - copyOnWrite:
        persistentVolumeClaim:
          claimName: my-cow-pvc
    name: pvc
```

I'll give this some thought. I think your way is the better way though right now.

**Comment by davidvossel** [comment]
I took @vladikr 's suggestion and restructured the api. Both the code and PR description have been updated to reflect this.

**Comment by mng-stark** [comment]
Why was this abandoned? This feature would greatly help in reducing storage footprint of VMs

**Comment by iholder101** [comment]
> Why was this abandoned? This feature would greatly help in reducing storage footprint of VMs

That's a good question.
This is a very old PR, hence I fear that your message won't be seen by the relevant people.

If you're interested in this, I'd suggest dropping a message on our [slack channel](https://kubernetes.slack.com/archives/C0163DT0R8X) or [mailing list](mailto:kubevirt-dev@googlegroups.com).

**Comment by mng-stark** [comment]
> > Why was this abandoned? This feature would greatly help in reducing storage footprint of VMs
> 
> That's a good question. This is a very old PR, hence I fear that your message won't be seen by the relevant people.
> 
> If you're interested in this, I'd suggest dropping a message on our [slack channel](https://kubernetes.slack.com/archives/C0163DT0R8X) or [mailing list](mailto:kubevirt-dev@googlegroups.com).

Thank you for your reply. I have been able to rebase the PRs functionality (with additional snapshotting support) onto stable 1.6. in my fork. However I do not have the knowledge nor the time to perform a proper PR procedure here to propose my changes. Therefore I'm offering my changes, contained in https://github.com/mng-stark/kubevirt/tree/feature/delta-cow to the public domain.

**Comment by iholder101** [comment]
> > > Why was this abandoned? This feature would greatly help in reducing storage footprint of VMs
> > 
> > 
> > That's a good question. This is a very old PR, hence I fear that your message won't be seen by the relevant people.
> > If you're interested in this, I'd suggest dropping a message on our [slack channel](https://kubernetes.slack.com/archives/C0163DT0R8X) or [mailing list](mailto:kubevirt-dev@googlegroups.com).
> 
> Thank you for your reply. I have been able to rebase the PRs functionality (with additional snapshotting support) onto stable 1.6. in my fork. However I do not have the knowledge nor the time to perform a proper PR procedure here to propose my changes. Therefore I'm offering my changes, contained in [mng-stark/kubevirt@`feature`/delta-cow](https://github.com/mng-stark/kubevirt/tree/feature/delta-cow?rgh-link-date=2025-11-17T15%3A19%3A09Z) to the public domain.

Thank you very much.
If you have the time, I think it would be beneficial to open an upstream issue or drop a message in our slack / mailing list. This way, even if you don't have the time to merge these changes, there's a chance that someone else will be motivated to do so.

Thanks in advance! :)

**Comment by mng-stark** [comment]
I've dropped a message in the mailing list.

**Pull request closed without merging**